### ISSUE TITLE
KafkaConsumer close is blocking forever

### ISSUE THREAD

**Issue by zkaiwen** [edenhill/librdkafka#1002]
Description
===========
One of my consumers during close is blocking forever and not exiting. I see that the consumer should be doing rebalancing, stopping consuming, and commiting offsets from the API, but i'm not sure how to go about debugging where the issue is or to see why it might be hanging. Is there any suggestion as to go about or see why the KafkaConsumer::close would be hanging. Thanks

How to reproduce
================

Checklist
=========
Please provide the following information:

 - [x] librdkafka version (release number or git tag): 0.9.2
 - [x] Apache Kafka version: 0.10
 - [ ] librdkafka client configuration:
 - [x] Operating system: centos7
 - [ ] Using the legacy Consumer
 - [x] Using the high-level KafkaConsumer
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
If brokers are unavailable close() might hang for socket.timeout.ms (60s) trying to commit the final offsets, but it sounds like it hangs longer than that.
Try reproducing this with `debug=cgrp,topic,protocol` and provide me with the logs and I'll see what I can find.

**Comment by zkaiwen** [comment]
This is the output when the close function is called. 

[INDEXER 2] -- CLosing consumer
%7|1484596276.598|RECV|rdkafka#consumer-1| kafka:9092/1001: Received FetchResponse (v1, 40 bytes, CorrId 441, rtt 99.73ms)
%7|1484596276.598|SEND|rdkafka#consumer-1| kafka:9092/1001: Sent FetchRequest (v1, 67 bytes @ 0, CorrId 442)
%7|1484596276.698|RECV|rdkafka#consumer-1| kafka:9092/1001: Received FetchResponse (v1, 40 bytes, CorrId 442, rtt 99.73ms)
%7|1484596276.698|SEND|rdkafka#consumer-1| kafka:9092/1001: Sent FetchRequest (v1, 67 bytes @ 0, CorrId 443)
%7|1484596276.798|RECV|rdkafka#consumer-1| kafka:9092/1001: Received FetchResponse (v1, 40 bytes, CorrId 443, rtt 99.73ms)
%7|1484596276.798|SEND|rdkafka#consumer-1| kafka:9092/1001: Sent FetchRequest (v1, 67 bytes @ 0, CorrId 444)
%7|1484596276.824|CGRPOP|rdkafka#consumer-1| Group "testgroup" received op TERMINATE (v0) in state up (join state started, v8 vs 0)
%7|1484596276.824|CGRPTERM|rdkafka#consumer-1| Terminating group "testgroup" in state up with 1 partition(s)
%7|1484596276.824|UNSUBSCRIBE|rdkafka#consumer-1| Group "testgroup": unsubscribe from current subscription of 1 topics (leave group=yes, join state started, v8)
%7|1484596276.824|SUBSCRIPTION|rdkafka#consumer-1| Group "testgroup": clearing subscribed topics list (1)
%7|1484596276.824|SUBSCRIPTION|rdkafka#consumer-1| Group "testgroup": effective subscription list changed from 1 to 0 topic(s):
%7|1484596276.824|PAUSE|rdkafka#consumer-1| Library pausing 1 partition(s)
%7|1484596276.824|BARRIER|rdkafka#consumer-1| testSigQ [0]: rd_kafka_toppar_op_pause_resume:2051: new version barrier v7
%7|1484596276.824|PAUSE|rdkafka#consumer-1| Pause testSigQ [0] (v7)
%7|1484596276.824|CGRPJOINSTATE|rdkafka#consumer-1| Group "testgroup" changed join state started -> wait-unassign (v8, state up)
%7|1484596276.824|BARRIER|rdkafka#consumer-1| Group "testgroup": rd_kafka_cgrp_unassign:1471: new version barrier v9
%7|1484596276.824|UNASSIGN|rdkafka#consumer-1| Group "testgroup": unassigning 1 partition(s) (v9)
%7|1484596276.824|OFFSET|rdkafka#consumer-1| Topic testSigQ [0]: stored offset 670, committed offset 670
%7|1484596276.824|OFFSET|rdkafka#consumer-1| Topic testSigQ [0]: setting offset INVALID for commit
%7|1484596276.824|COMMIT|rdkafka#consumer-1| OffsetCommit internal error: Local: No offset stored
%7|1484596276.824|COMMIT|rdkafka#consumer-1| OffsetCommit for 1 partition(s) returned: Local: No offset stored
%7|1484596276.824|UNASSIGN|rdkafka#consumer-1| Unassign not done yet (0 wait_unassign, 1 assigned, 0 wait commit): OffsetCommit done (__NO_OFFSET)
%7|1484596276.824|BARRIER|rdkafka#consumer-1| testSigQ [0]: rd_kafka_toppar_op_fetch_stop:1993: new version barrier v8
%7|1484596276.824|CONSUMER|rdkafka#consumer-1| Stop consuming testSigQ [0] (v8)
%7|1484596276.824|DESP|rdkafka#consumer-1| Removing (un)desired topic testSigQ [0]
%7|1484596276.824|RESUME|rdkafka#consumer-1| Library resuming 1 partition(s)
%7|1484596276.824|BARRIER|rdkafka#consumer-1| testSigQ [0]: rd_kafka_toppar_op_pause_resume:2051: new version barrier v9
%7|1484596276.824|RESUME|rdkafka#consumer-1| Resume testSigQ [0] (v9)
%7|1484596276.824|UNASSIGN|rdkafka#consumer-1| Unassign not done yet (1 wait_unassign, 1 assigned, 0 wait commit): unassign
%7|1484596276.824|CGRPTERM|rdkafka#consumer-1| Group "testgroup": waiting for 1 toppar(s), 1 unassignment(s), 0 commit(s) (state up, join-state wait-unassign) before terminating
%7|1484596276.824|CGRPTERM|rdkafka#consumer-1| Group "testgroup": waiting for 1 toppar(s), 1 unassignment(s), 0 commit(s) (state up, join-state wait-unassign) before terminating
%7|1484596276.824|OP|rdkafka#consumer-1| testSigQ [0] received op PAUSE (v7) in fetch-state active (opv6)
%7|1484596276.824|PAUSE|rdkafka#consumer-1| Pause testSigQ [0]: at offset 670 (state active, v7)
%7|1484596276.824|OP|rdkafka#consumer-1| testSigQ [0] received op FETCH_STOP (v8) in fetch-state active (opv7)
%7|1484596276.824|FETCH|rdkafka#consumer-1| Stopping fetch for testSigQ [0] in state active (v8)
%7|1484596276.825|PARTSTATE|rdkafka#consumer-1| Partition testSigQ [0] changed fetch state active -> stopping
%7|1484596276.825|STORETERM|rdkafka#consumer-1| testSigQ [0]: offset store terminating
%7|1484596276.825|PARTSTATE|rdkafka#consumer-1| Partition testSigQ [0] changed fetch state stopping -> stopped
%7|1484596276.825|OP|rdkafka#consumer-1| testSigQ [0] received op PAUSE (v9) in fetch-state stopped (opv8)
%7|1484596276.825|RESUME|rdkafka#consumer-1| Not resuming stopped testSigQ [0]: at offset 670 (state stopped, v9)
%7|1484596276.825|CGRPOP|rdkafka#consumer-1| Group "testgroup" received op PARTITION_LEAVE in state up (join state wait-unassign, v9) for testSigQ [0]
%7|1484596276.825|PARTDEL|rdkafka#consumer-1| Group "testgroup": delete testSigQ [0]
%7|1484596276.825|CGRPTERM|rdkafka#consumer-1| Group "testgroup": waiting for 0 toppar(s), 1 unassignment(s), 0 commit(s) (state up, join-state wait-unassign) before terminating
%7|1484596276.825|CGRPOP|rdkafka#consumer-1| Group "testgroup" received op REPLY:FETCH_STOP in state up (join state wait-unassign, v9) for testSigQ [0]
%7|1484596276.825|UNASSIGN|rdkafka#consumer-1| Group "testgroup": unassign done in state up (join state wait-unassign): without new assignment: FETCH_STOP done
%7|1484596276.825|LEAVE|rdkafka#consumer-1| Group "testgroup": leave
%7|1484596276.825|CGRPJOINSTATE|rdkafka#consumer-1| Group "testgroup" changed join state wait-unassign -> init (v9, state up)
%7|1484596276.825|CGRPSTATE|rdkafka#consumer-1| Group "testgroup" changed state up -> term (v9, join-state init)
%7|1484596276.825|FETCHADD|rdkafka#consumer-1| kafka:9092/1001: Removed testSigQ [0] from fetch list (0 entries, opv 6)
%7|1484596276.825|SEND|rdkafka#consumer-1| kafka:9092/1001: Sent LeaveGroupRequest (v0, 78 bytes @ 0, CorrId 445)
%7|1484596276.825|BRKUNASSIGN|rdkafka#consumer-1| Group "testgroup" management unassigned from broker handle kafka:9092/1001
%7|1484596276.825|DESTROY|rdkafka#consumer-1| Terminating instance
%7|1484596276.825|DESTROY|rdkafka#consumer-1| Destroy internal
%7|1484596276.825|DESTROY|rdkafka#consumer-1| Remove all topics
%7|1484596276.825|PARTCNT|rdkafka#consumer-1| Topic testSigQ partition count changed from 1 to 0
%7|1484596276.825|REMOVE|rdkafka#consumer-1| testSigQ [0] no longer reported in metadata
%7|1484596276.825|BRKMIGR|rdkafka#consumer-1| testSigQ [0] 0x7fece8003ca0 sending final LEAVE for removal by kafka:9092/1001
%7|1484596276.825|TOPPARREMOVE|rdkafka#consumer-1| Removing toppar testSigQ [-1] 0x7fece8003820
%7|1484596276.825|DESTROY|rdkafka#consumer-1| testSigQ [-1]: 0x7fece8003820 DESTROY_FINAL
%7|1484596276.825|TOPBRK|rdkafka#consumer-1| kafka:9092/1001: Topic testSigQ [0]: leaving broker (0 messages in xmitq, next leader (none), rktp 0x7fece8003ca0)
%7|1484596276.825|TOPBRK|rdkafka#consumer-1| kafka:9092/1001: Topic testSigQ [0]: no next leader, failing 0 message(s) in partition queue
%7|1484596276.827|BROKERFAIL|rdkafka#consumer-1| :0/internal: failed: err: Local: Broker handle destroyed: (errno: Success)
%7|1484596276.898|RECV|rdkafka#consumer-1| kafka:9092/1001: Received FetchResponse (v1, 40 bytes, CorrId 444, rtt 99.63ms)
%7|1484596276.898|UNKTOPIC|rdkafka#consumer-1| kafka:9092/1001: Received Fetch response (error 0) for unknown topic testSigQ [0]: ignoring
%7|1484596276.898|RECV|rdkafka#consumer-1| kafka:9092/1001: Received LeaveGroupResponse (v0, 2 bytes, CorrId 445, rtt 73.36ms)

**Comment by bjustin-ibm** [comment]
@edenhill I am also running into a very similar issue. Here is the output of calling `close()` with the debug flags you mentioned above:

```
%7|1484688543.314|CGRPOP|rdkafka#consumer-1| Group "/bjustin@us.ibm.com_demo/TestKafkaTrigger" received op TERMINATE (v0) in state up (join state wait-assign-rebalance_cb, v1)
%7|1484688543.314|CGRPTERM|rdkafka#consumer-1| Terminating group "/bjustin@us.ibm.com_demo/TestKafkaTrigger" in state up with 0 partition(s)
%7|1484688543.314|BARRIER|rdkafka#consumer-1| Group "/bjustin@us.ibm.com_demo/TestKafkaTrigger": rd_kafka_cgrp_assign:1482: new version barrier v2
%7|1484688543.314|ASSIGN|rdkafka#consumer-1| Group "/bjustin@us.ibm.com_demo/TestKafkaTrigger": assigning 0 partition(s) in join state wait-assign-rebalance_cb
%7|1484688543.314|CGRPJOINSTATE|rdkafka#consumer-1| Group "/bjustin@us.ibm.com_demo/TestKafkaTrigger" changed join state wait-assign-rebalance_cb -> assigned (v2)
%7|1484688543.314|UNSUBSCRIBE|rdkafka#consumer-1| Group "/bjustin@us.ibm.com_demo/TestKafkaTrigger": unsubscribe from current subscription of 1 topics (leave group=yes, join state assigned, v2)
%7|1484688543.314|SUBSCRIPTION|rdkafka#consumer-1| Group "/bjustin@us.ibm.com_demo/TestKafkaTrigger": clearing subscribed topics list (1)
%7|1484688543.314|SUBSCRIPTION|rdkafka#consumer-1| Group "/bjustin@us.ibm.com_demo/TestKafkaTrigger": effective subscription list changed from 1 to 0 topic(s):
%7|1484688543.314|UNASSIGN|rdkafka#consumer-1| Group "/bjustin@us.ibm.com_demo/TestKafkaTrigger": unassign done in state up (join state assigned): without new assignment
%7|1484688543.314|LEAVE|rdkafka#consumer-1| Group "/bjustin@us.ibm.com_demo/TestKafkaTrigger": leave
%7|1484688543.314|CGRPJOINSTATE|rdkafka#consumer-1| Group "/bjustin@us.ibm.com_demo/TestKafkaTrigger" changed join state assigned -> init (v2)
%7|1484688543.314|CGRPSTATE|rdkafka#consumer-1| Group "/bjustin@us.ibm.com_demo/TestKafkaTrigger" changed state up -> term (v2)
%7|1484688543.314|BRKUNASSIGN|rdkafka#consumer-1| Group "/bjustin@us.ibm.com_demo/TestKafkaTrigger" management unassigned from broker handle sasl_ssl://kafka04-prod01.messagehub.services.us-south.bluemix.net:9093/3
%7|1484688543.318|SEND|rdkafka#consumer-1| sasl_ssl://kafka04-prod01.messagehub.services.us-south.bluemix.net:9093/3: Sent LeaveGroupRequest (v0, 110 bytes @ 0, CorrId 6)
%7|1484688543.351|RECV|rdkafka#consumer-1| sasl_ssl://kafka04-prod01.messagehub.services.us-south.bluemix.net:9093/3: Received HeartbeatResponse (v0, 2 bytes, CorrId 5, rtt 135.09ms)
%7|1484688543.485|RECV|rdkafka#consumer-1| sasl_ssl://kafka04-prod01.messagehub.services.us-south.bluemix.net:9093/3: Received LeaveGroupResponse (v0, 2 bytes, CorrId 6, rtt 167.10ms)
```

Incidentally, all threads in my multi-threaded app seems to hang when any thread calls `close()`. Should I open a separate issue for that?

**Comment by edenhill** [comment]
This is most likely fixed on master (soon 0.9.4)

**Comment by hrjaco** [comment]
@edenhill I am using librdkafka 0.9.5 and I still see this forever hanging issue occasionally after calling close(). Is there any other case that might cause this issue?
